lib/orangelight/browse_lists/call_number_csv.rb

```diff
@@ -0,0 +1,131 @@
+# frozen_string_literal: true
+
+module BrowseLists
+  class SolrResponseError < StandardError; end
+  class CallNumberCSV
+    attr_reader :output_root, :facet_request, :conn, :rows
+    attr_accessor :multi_cn_lookup
+    def initialize(facet_request, conn, output_root: Pathname.new("/tmp"), rows: 5_000)
+      @output_root = output_root
+      @facet_request = facet_request
+      @conn = conn
+      @rows = rows
+      @multi_cn_lookup = {}
+    end
+
+    def write
+      write_multiple_call_numbers
+      write_remaining_call_numbers
+    end
+
+    def filename
+      output_root.join("#{facet_field}.csv")
+    end
+
+    private
+
+      def core_url
+        Blacklight.default_index.connection.uri.to_s.gsub(%r{^.*\/solr}, '/solr')
+      end
+
+      def facet_field
+        "call_number_browse_s"
+      end
+
+      # Collect all the call numbers with facet count > 2
+      # Write each as a row in the csv and add it to the lookup table
+      def write_multiple_call_numbers
+        # This request takes a minute or so
+        resp = conn.get "#{facet_request}#{facet_field}&facet.mincount=2"
+        req = JSON.parse(resp.body)
+        CSV.open(filename, 'wb') do |csv|
+          req['facet_counts']['facet_fields'][facet_field].each_slice(2) do |mcn, record_count|
+            sort_cn = StringFunctions.cn_normalize(mcn)
+            multi_cn_lookup[sort_cn] = record_count
+            csv << [sort_cn, mcn, 'ltr', '', "#{record_count} titles with this call number", '', '', "?f[#{facet_field}][]=#{CGI.escape(mcn)}", '', 'Multiple locations']
+          end
+        end
+      end
+
+      # Append the rest of the call numbers to the file
+      def write_remaining_call_numbers
+        cursor_mark = '*'
+        CSV.open(filename, 'ab') do |csv|
+          loop do
+            body = solr_cursor_page_body(cursor_mark)
+            body['response']['docs'].each do |record|
+              next unless record[facet_field]
+              record[facet_field].each do |cn|
+                sort_cn = StringFunctions.cn_normalize(cn)
+                next if multi_cn_lookup.key?(sort_cn)
+                csv << parse_cn_row(record, cn, sort_cn)
+              end
+            end
+            next_cursor_mark = body['nextCursorMark']
+            break if cursor_mark == next_cursor_mark
+            cursor_mark = next_cursor_mark
+          end
+        end
+      end
+
+      # Get a page of solr results for actual items starting at the given number
+      def solr_cursor_page_body(cursor_mark)
```

Comment:
question: Is the cursor mark a number? Curious about the comment here. Maybe a YARD-doc would help here, if it isn't refactored.

Reply:
Oh this comment is wrong since it was changed from using the start number. thanks.

Reply:
oh, and it's not a number it's an encoded string. example: `AoEoMTAwMDA1NjY=`